prow/pod-utils/gcs/upload.go

```diff
@@ -171,12 +197,51 @@ func (w gcsObjectWriter) ApplyAttributes(attrs *storage.ObjectAttrs) {
 	w.Writer.ObjectAttrs = *attrs
 }
 
+type openerObjectWriter struct {
+	pkgio.Opener
+	Bucket      string
+	Dest        string
+	writeCloser pkgio.WriteCloser
+}
+
+func (w *openerObjectWriter) Write(p []byte) (n int, err error) {
+	if w.writeCloser == nil {
+		w.writeCloser, err = w.Opener.Writer(context.Background(), fmt.Sprintf("%s/%s", w.Bucket, w.Dest))
+		if err != nil {
+			return 0, err
+		}
+	}
+	return w.writeCloser.Write(p)
+}
+
+func (w *openerObjectWriter) Close() error {
+	if w.writeCloser != nil {
+		err := w.writeCloser.Close()
+		w.writeCloser = nil
+		return err
+	}
+	return nil
+}
+
+func (w *openerObjectWriter) SetSize(_ int64) {
+	// ChunkSize is in gocloud only implemented by GCS
+	// but we're not using the openerObjectWriter for GCS
+	// so we don't have to set this
+	return
+}
+
+func (w *openerObjectWriter) ApplyAttributes(_ *storage.ObjectAttrs) {
```

Comment:
if we ever want to use this for GCS as well or pass any kind of opts, we need to change the `Writer` method in the `io` package to accept some kind of opts. Could probably be a simplistic struct that only contains GCS options and gets extended in the future if the need arises. We can do this backwards-compatible by leveraging the functional opts pattern.

Reply:
I resolve this as it's handled in: https://github.com/kubernetes/test-infra/pull/16722#discussion_r400704730